minetest/minetest
# [no squash] Extract image generation/cache from texturesource.cpp and some refactoring

## Goal of the PR
- This is the followup PR of #14401
- The goal is to improve `texturesource.cpp` such that it only handles texture storage, and no source image caching or applying of texture modifiers.

## How does the PR work?
- Everything related to image generation/modification and source images caching moved to `imagesource.h/cpp`
- Since those are the parts which can only be accessed from the main tread, it simplifies the tread safety code in `texturesource.cpp`, and `imagesource.cpp` doesn't need any mutex.
- The remaining `texturesource.cpp` gets refactored, but it's mostly modernizing and formatting, so there shouldn't be much behavioral difference.
- `SourceImageCache` does also get refactored a little bit, but the rest of  `imagesource.cpp` stays the same, since it would be too much for this PR to also reactor it.
- Information to hopefully make it easier to review:
  - `IImage` is an image object, which can be modified.
  - "source image" is an `IImage` loaded from a texture file without any modification.
  - `ITexture` is made from an `IImage` for the driver to use. It doesn't get modified.



## Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
2.2 Internal code refactoring

## To do

This PR is Ready for Review.


(The next step is to refactor `imagesource.cpp`.)
(In the future, it may be good to replace const string reverences by string views everywhere, but this would involve changing many files.)

## How to test
- See that the functions which moved to `imagesource.cpp` didn't change.
  (e.g. use `git diff --color-moved <commit1> <commit2>`)
- Read the refactored code.
